presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/FunctionCallProvider.java

```diff
@@ -98,12 +100,28 @@ public boolean equals(Object object)
             FunctionCall other = (FunctionCall) object;
 
             return Objects.equals(name, other.getName()) &&
-                    other.getWindow().isPresent() &&
-                    Objects.equals(frame, other.getWindow().get().getFrame()) &&
+                    matchWindowFrame(other.getWindow()) &&
                     Objects.equals(distinct, other.isDistinct()) &&
                     Objects.equals(getArguments(), other.getArguments());
         }
 
+        private boolean matchWindowFrame(Optional<Window> other)
```

Comment:
Good lord both of you!

1. Use DeMorgan's Laws
2. Get rid of the negation in the variable name. It's hard to reason about.
3. One has a frame and the other doesn't can be tested with !=
4. This shouldn't be so complicated it needs tests of its own

```
        private boolean matchWindowFrame(Optional<Window> other)
        {
            boolean otherFramePresent = other.isPresent()
                    && (other.get() instanceof WindowInline)
                    && (((WindowInline) other.get()).getFrame().isPresent());

            // One has a frame and the other does not.
            if (frame.isPresent() != otherFramePresent) {
                return false;
            }

            // Neither has a frame
            if (!frame.isPresent() && !otherFramePresent) {
                return false;
            }

            return Objects.equals(frame, ((WindowInline) other.get()).getFrame());
        }
```

More to the point, if we're comparing the expected to actual with `EWFC.equals()`, just drop the check on the Window.Frame entirely from `.equals()`. Per @maciejgrzybek, we don't use that information for anything past this point, and there's no real need to validate it other than in WindowNode.Frame.

Apologies for misleading @amrutagokhale [here](https://github.com/prestodb/presto/pull/8307#issuecomment-323762755):
> For the sake of posterity, the other obvious approach of changing to only verifying the WindowNode.Frame information and ignoring the FunctionCall frame information is problematic because we compare FunctionCalls with equality, so they either need to be equal (as proposed above), or we need to implement and equals-except-frame function, which leaves a hidden maintenance touch-point for somebody in the future.

It turns out we already have the hidden maintenance touch-point (which I forgot about), so we may as well just change it to do what we actually care about.

Reply:
@ebd2 I replaced the code with your version :) I need to think a bit more about dropping this check entirely from `EWFC.equals()`.

Reply:
If we are dropping the check on WindowFrame from `EWFC.equals()`, we are essentially choosing to verify the frame information in the plan using WindowNode.Frame instead of WindowFrame. So we can remove the `Optional<WindowFrame>` altogether from `FunctionCallProvider`. I have pushed a commit making such a change. Actually, we might be able to get rid of `isWindowFunction` member and remove the `EWFC` class as well.

Reply:
I'm reasonably certain you still need EWFC (or something largely equivalent implementing its .equals() method) because we can't use FC.equals to do the expected/actual comparison because we don't create a full FC for the expected value.

But I've been wrong before ;-)

Reply:
Specifically, for the case where the FC *does* contain a window, we don't want to have to worry about populating it correctly to make the comparison work if we only care about validating `WindowNode.frame`. Which is how we got here in the first place, I think.

Reply:
Ah, I always forget the fact that `EWFC` extends FC rather than FunctionCallProvider (because it is present in FunctionCallProvider.java). I see your point about `EWFC.equals()` not being equivalent to `FC.equals()` and that's how we would like to keep it, given that we are not comparing window related information when comparing two `EWFC` objects. I am just going to add a comment in `EWFC.equals()` about this.